### PR TITLE
Implement Game Levels with Win and Lose Conditions

### PR DESCRIPTION
**Pull request by neoimpulse**
- Design and Implemented game levels with with and lose conditions.
- Added GameManager.cs to handle win-lose conditions. Updated UIManager.cs and ShapesManager.cs.

### CODE REVIEW

File: Assets/Scripts/GameManager.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class GameManager : MonoBehaviour
+{
+    private float _currentTime = 0;
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs]: i tend to prefer putting private variables that aren't serialized down below the serialized ones and with other non-serialized private variables.

File: Assets/Scripts/GameManager.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    [SerializeField]
+    private float startingTime = 60f;
+    [SerializeField]
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs:10]: preference for putting new lines between a variable and the next attribute

```suggestion

    [SerializeField]
```

File: Assets/Scripts/GameManager.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    private int currentLevel = 1;
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs]: naming should be consistent -- if you're going to preface private vars with an underscore, they should all be that way. (i tend to underscore all non-serialized private vars but not underscore the serialized ones -- i'm sure other unity devs would disagree with that practice. folks tend to value consistency)

File: Assets/Scripts/GameManager.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    private UIManager _UIManager;
+    private ShapesManager _shapesManager;
+
+    // Start is called before the first frame update
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs]: you can remove these template comments from Unity, they don't add any value

File: Assets/Scripts/GameManager.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    void Start()
+    {
+        _UIManager = GameObject.Find("Canvas").GetComponent<UIManager>();
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs]: So using `.Find()` is a *highly* discouraged practice, especially in this case. There are multiple much better ways you can approach this that are way more performant than `.Find()` and `.GetComponent()`.

1) Add the `[SerializedField]` attribute to `_UIManager` and `_ShapesManager`. Then, the game objects that have those scripts on them can be dragged into the field in the editor, giving you direct access to those instances in this script.

2) You can make `UIManager` and `ShapesManager` [global singletons](https://gamedevbeginner.com/singletons-in-unity-the-right-way/) that are accessible via a call like `UIManager.instance`. This way you wouldn't even have to link the game objects, you could just call the instance from anywhere in the code. NOTE: the singleton route should be used in cases where only 1 of that class should ever exist throughout the lifetime of the game. (That said, you could potentially make a case for both `UIManager` and `ShapesManager` being viable singleton candidates in this context. Up to you.)

File: Assets/Scripts/GameManager.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        _shapesManager = GameObject.Find("ShapesManager").GetComponent<ShapesManager>();
+        _currentTime = startingTime;
+
+    }
+
+    // Update is called once per frame
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs]: you can remove this.

File: Assets/Scripts/GameManager.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        _currentTime -= 1 * Time.deltaTime;
+        _UIManager.CountdownTimerText(_currentTime);
+
+        if (_currentTime <= 0)
+        {
+            _currentTime = 0;
+            LoseState();
+        }
+    }
+
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs:43]: unnecessary extra newline here

File: Assets/Scripts/GameManager.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    public void LoseState()
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs]: does this method need to be public? does any other class need access to it?

File: Assets/Scripts/GameManager.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        _UIManager.PlayerLoseSequence();
+        Time.timeScale = 0;
+    }
+
+    public void WinState()
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs:50]: same here. does this method need to be public? does any other class need access to it?

File: Assets/Scripts/GameManager.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    {        
+        _UIManager.PlayerWinsSequence();
+        Time.timeScale = 0;
+    }
+
+    public void NextLevel()
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs]: and here. does this method need to be public? does any other class need access to it?

File: Assets/Scripts/GameManager.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        _UIManager.LevelText(currentLevel);
+        _currentTime = startingTime;
+        _UIManager.NextLevelUIReset();
+        _shapesManager.InitializeTypesOnPrefabShapesAndBonuses();
+        _shapesManager.InitializeCandyAndSpawnPositions();
+        _shapesManager.StartCheckForPotentialMatches();
+        _shapesManager.InitializeVariables();
+    }
+
+    public void RestartLevel()
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs]: this one too - does this method need to be public? does any other class need access to it?

File: Assets/Scripts/GameManager.cs [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        Time.timeScale = 1;
+        _currentTime = startingTime;
+        _UIManager.RestartLevelUIReset();
```
Comment (dalarose) [inline Assets/Scripts/GameManager.cs:75]: There's a lot of duplicated code here between `NextLevel` and `RestartLevel` that could be refactored. I would put all of the duplicated code in a private helper function that both `NextLevel` and `RestartLevel` call.

File: Assets/Scripts/ShapesManager.cs [hunk 13]
```diff
@@ -53,14 +54,15 @@ void Start()
         {
             Debug.LogError("UI Manager is NUll");
         }
+        _gameManager = GameObject.Find("GameManager").GetComponent<GameManager>();
```
Comment (dalarose) [inline Assets/Scripts/ShapesManager.cs]: Yeah....no more `.Find()`s please. 😬

File: Assets/Scripts/UIManager.cs [hunk 14]
```diff
@@ -7,11 +7,25 @@ public class UIManager : MonoBehaviour
 {
     [SerializeField]
     private Text _winText;
+    [SerializeField]
```
Comment (dalarose) [inline Assets/Scripts/UIManager.cs:11]: preference for spaces between variable and the next `[SerializedField]` attribute

File: Assets/Scripts/UIManager.cs [hunk 15]
```diff
@@ -7,11 +7,25 @@ public class UIManager : MonoBehaviour
...
+    private Text _LevelText;
+    [SerializeField]
+    private Text _timerText;
+    [SerializeField]
+    private Button _nextLevelButton;
+    [SerializeField]
+    private Button _restartLevelButton;
 
     // Start is called before the first frame update
```
Comment (dalarose) [inline Assets/Scripts/UIManager.cs]: can remove this.

File: Assets/Scripts/UIManager.cs [hunk 16]
```diff
@@ -7,11 +7,25 @@ public class UIManager : MonoBehaviour
...
     void Start()
     {
         _winText.gameObject.SetActive(false);
+        _loseText.gameObject.SetActive(false);
+        _nextLevelButton.gameObject.SetActive(false);
+        _restartLevelButton.gameObject.SetActive(false);
+        
     }
 
     // Update is called once per frame
```
Comment (dalarose) [inline Assets/Scripts/UIManager.cs]: can remove this.

File: Assets/Scripts/UIManager.cs [hunk 17]
```diff
@@ -7,11 +7,25 @@ public class UIManager : MonoBehaviour
 {
     [SerializeField]
     private Text _winText;
+    [SerializeField]
+    private Text _loseText;
+    [SerializeField]
+    private Text _LevelText;
```
Comment (dalarose) [inline Assets/Scripts/UIManager.cs]: this variable starts with a capital letter but the others don't

File: Assets/Scripts/GameManager.cs (hunk 7 above)
Comment (neoimpulse) [reply Assets/Scripts/GameManager.cs:42]: Was this in reference to Line 42? or 35?

File: Assets/Scripts/GameManager.cs (hunk 9 above)
Comment (neoimpulse) [reply Assets/Scripts/GameManager.cs:50]: For this method, another class in ShapesManager needs access to it.

**Review by dalarose — APPROVED** [review]

**Pull request merged**